### PR TITLE
gyration_moments method / allow shape parameters to yield per residue quantities

### LINKED ISSUE #2292: Add "unwrap" keyword to shape_parameter
**Is your feature request related to a problem? Please describe.**
"unwrap" keyword needs to be added to the shape_parameter method

### LINKED ISSUE #3002: Some methods that accept the compound keyword fail or don't return the expected set of values
For at least `moment_of_inertia` and `asphericity` the use of compounds seems to be broken:

## Expected behavior ##
```python
from MDAnalysis.tests.datafiles import TRZ_psf, TRZ
u = mda.Universe(TRZ_psf, TRZ) # universe with 24 residues

u.atoms.asphericity(compound='residues')
# expected a 1D array with 24 values of asphericity, one per residue
# equivalent to
np.array([res.atoms.asphericity() for res in u.residues])

u.atoms.moment_of_inertia(compound='residues')
# expected a 3D array with 24 MoI 3x3 tensors, one per residue, with shape (24,3,3)
# equivalent to
np.array([res.atoms.moment_of_inertia() for res in u.residues])
```
## Actual behavior ##

```python
u.atoms.asphericity(compound='residues')
~/src/MDAnalysis/mdanalysis/package/MDAnalysis/lib/util.py in wrapper(group, *args, **kwargs)
   1841         # has already been thrown:
   1842         if group.isunique or warn_if_not_unique.warned:
-> 1843             return groupmethod(group, *args, **kwargs)
   1844         # Otherwise, throw a DuplicateWarning and execute the method.
   1845         method_name = ".".join(

~/src/MDAnalysis/mdanalysis/package/MDAnalysis/core/groups.py in wrapped(group, *args, **kwargs)
    386                 raise ValueError(
    387                     "both 'pbc' and 'unwrap' can not be set to true")
--> 388         return function(group, *args, **kwargs)
    389     return wrapped
    390

~/src/MDAnalysis/mdanalysis/package/MDAnalysis/core/topologyattrs.py in asphericity(group, pbc, unwrap, compound)
   1418             pbc=pbc, unwrap=unwrap, compound=compound)
   1419         if compound != 'group':
-> 1420             com = (com * group.masses[:, None]
   1421                    ).sum(axis=0) / group.masses.sum()
   1422

ValueError: operands could not be broadcast together with shapes (24,3) (8184,1)
```

```python
u.atoms.moment_of_inertia(compound='residues')
~/src/MDAnalysis/mdanalysis/package/MDAnalysis/lib/util.py in wrapper(group, *args, **kwargs)
   1841         # has already been thrown:
   1842         if group.isunique or warn_if_not_unique.warned:
-> 1843             return groupmethod(group, *args, **kwargs)
   1844         # Otherwise, throw a DuplicateWarning and execute the method.
   1845         method_name = ".".join(

~/src/MDAnalysis/mdanalysis/package/MDAnalysis/core/groups.py in wrapped(group, *args, **kwargs)
    386                 raise ValueError(
    387                     "both 'pbc' and 'unwrap' can not be set to true")
--> 388         return function(group, *args, **kwargs)
    389     return wrapped
    390

~/src/MDAnalysis/mdanalysis/package/MDAnalysis/core/topologyattrs.py in moment_of_inertia(group, pbc, **kwargs)
   1260             pbc=pbc, unwrap=unwrap, compound=compound)
   1261         if compound != 'group':
-> 1262             com = (com * group.masses[:, None]
   1263                    ).sum(axis=0) / group.masses.sum()
   1264

ValueError: operands could not be broadcast together with shapes (24,3) (8184,1)
```

Calling the method on a `ResidueGroup` seems to default to using `compound='group'` and yield a single value. More puzzingly, calling it on a `ResidueGroup` and explicitly passing `compound='residues'` works (no `ValueError`) but yields the same single value (with some precision difference betraying a different logic route):

```python
u.residues.asphericity()
>> 0.5655677527570107

u.residues.asphericity(compound='residues')
>> 0.5655677527570102
```

```python
u.residues.moment_of_inertia()
>> array([[1332923.85838838,  480989.6752955 , -227811.7564208 ],
          [ 480989.6752955 ,  612813.7537505 ,  231931.75813526],
          [-227811.7564208 ,  231931.75813526, 1509650.41986839]])

u.residues.moment_of_inertia(compound='residues')
>> array([[1332923.85838838,  480989.6752955 , -227811.7564208 ],
          [ 480989.6752955 ,  612813.7537505 ,  231931.75813526],
          [-227811.7564208 ,  231931.75813526, 1509650.41986839]])```
```

A consolidated compound-splitting strategy (#3000) might help fix these issues and uniformize the API.

## Current version of MDAnalysis ##

- Which version are you using? `2.0.0-dev0` (at commit beb5232dc7501e6bc0138036f697aa14da2a4162)
- Which version of Python? `3.6.9`
- Which operating system? Ubuntu 18.04

### LINKED ISSUE #3904: Retrieve per residue moments of the gyration tensor
## Is your feature request related to a problem? ##
It would be nice to retrieve the moments of the gyration tensor for custom shape assessments.

### PR DESCRIPTION
**Pull request by jaclark5**
… per residue quantities

Fixes #3002 
Fixes #3904 

Changes made in this Pull Request:
 - Added AtomGroup Method to provide group and per residue moments of gyration.

PR Checklist
------------
 - [X] Tests?
 - [X] Docs?
 - [X] CHANGELOG updated?
 - [X] Issue raised/referenced?

### CODE REVIEW

File: package/MDAnalysis/core/topologyattrs.py
Comment (hmacdope) [inline package/MDAnalysis/core/topologyattrs.py]: Could you please provide an equation and paper describing the property? I and users would benefit from more detailed docs. :)

File: package/MDAnalysis/core/topologyattrs.py
Comment (hmacdope) [inline package/MDAnalysis/core/topologyattrs.py]: What exception does this `try/except` cover and why? Naked try/excepts can be dangerous and I am sure what this one is supposed to guard?

File: testsuite/MDAnalysisTests/core/test_atomgroup.py
Comment (hmacdope) [inline testsuite/MDAnalysisTests/core/test_atomgroup.py:997]: Why has this gone from a scalar to an array?

File: testsuite/MDAnalysisTests/core/test_atomgroup.py
Comment (hmacdope) [inline testsuite/MDAnalysisTests/core/test_atomgroup.py:1013]: As above, why is this now an array vs scalar?

**Review by hmacdope — CHANGES_REQUESTED** [review]
Few comments and queries. :)

File: package/MDAnalysis/core/topologyattrs.py
Comment (jaclark5) [reply package/MDAnalysis/core/topologyattrs.py]: From what I understand, try/except statements are faster in python than testing with an if statement. If there is only one dimension the first option will fail. I changed it anyway.

File: testsuite/MDAnalysisTests/core/test_atomgroup.py
Comment (jaclark5) [reply testsuite/MDAnalysisTests/core/test_atomgroup.py:997]: As mentioned in https://github.com/MDAnalysis/mdanalysis/issues/3002, the shape and anisotropy functions produced one value no matter the chosen ``compound``. I implemented the coordinate handling use in the center of mass calculation (to handle ``pbc`` and ``wrap`` options with  the different ``compound ``options) in the new ``gyration_moments`` method which then also corrects this issue for these other metrics.

File: testsuite/MDAnalysisTests/core/test_atomgroup.py
Comment (jaclark5) [reply testsuite/MDAnalysisTests/core/test_atomgroup.py:1013]: Same as above

File: testsuite/MDAnalysisTests/core/test_atomgroup.py
Comment (hmacdope) [reply testsuite/MDAnalysisTests/core/test_atomgroup.py:997]: Is there a selection option for `compound` with the new setup that would recapitulate this test value?

File: testsuite/MDAnalysisTests/core/test_atomgroup.py
Comment (hmacdope) [reply testsuite/MDAnalysisTests/core/test_atomgroup.py:1013]: See response above :)

**Comment by hmacdope** [comment]
@jaclark5 would you be able to fix conflicts / trigger CI and we can go from there? Might also be good to have another pair of eyes @MDAnalysis/coredevs?

File: testsuite/MDAnalysisTests/core/test_atomgroup.py
Comment (jaclark5) [reply testsuite/MDAnalysisTests/core/test_atomgroup.py:997]: Based on my read of issue #3002 and an understanding of what specifying the `compound` keyword is supposed to achieve, I think the proposed change offers the expected and desired result. 
I could add the initial code back in with an additional option `cluster=False` to say that the shape of the group should be assessed with compounds unwrapped beforehand.

**Comment by jaclark5** [comment]
@hmacdope These errors occur from the new Bibtex citation style. This would be resolved with the following changes:
    
    maintainer/conda/environment.yml:  - pyyaml>=3.02
    package/requirements.txt:pyyaml>=3.02

Recall that these changes were added so that the tests would pass in my other PR #3842. So as soon as those are merged, I'll update this PR with the develop branch ;)

**Comment by jaclark5** [comment]
@hmacdope I'm now getting an error in the tests from MDAnalysis/coordinate/XTC.py

**Comment by IAlibay** [comment]
> I'm now getting an error in the tests from MDAnalysis/coordinate/XTC.py

@jaclark5 are you getting this locally? If you so you might need to pip install MDAnalysis again.

**Comment by jaclark5** [comment]
I am getting it locally. Good point, thanks.

Edit: That worked

File: testsuite/MDAnalysisTests/core/test_atomgroup.py
Comment (hmacdope) [reply testsuite/MDAnalysisTests/core/test_atomgroup.py:997]: Sorry I missed this comment. Would you mind explaining a little bit more what you mean? Cluster is not a prexisting arg right?

Comment (jaclark5) [reply testsuite/MDAnalysisTests/core/test_atomgroup.py:997]: Right, I'm suggesting that I add it if you want to retain the previous output behavior, but the asphericity was so far from expected behavior (#3002), unless I'm misunderstanding the code, I'm not sure it's worth it since the "correct" combination of the keywords (i.e., using compound="group".) should reproduce this single value result, and if compound equals anything else, then it's not expected/desired.

Old Asphericity steps:
1) Use `wrap`, `unwrap`, and `compound` to calculate an array/matrix of the center of masses
2) If compound != "group", combine the centers of mass into a single center of mass
    - Effectively obtaining the result for compound="group" but the expression multiplies the atomic mass (size (n_atoms,)) by the com array ((n_residues,3) or (n_segments,3)) and then sums them to produce a single COM (size (,3)), so I'm not sure how that dimension mismatch works.
3) Use single COM to calculate a single asphericity parameter

Unless I'm misunderstanding, steps 2 and 3 are not the desired behavior when compound != "group".

File: package/MDAnalysis/core/topologyattrs.py
Comment (hmacdope) [inline package/MDAnalysis/core/topologyattrs.py]:
```suggestion
        .. versionadded:: 2.5.0
```

File: package/MDAnalysis/core/topologyattrs.py
Comment (hmacdope) [inline package/MDAnalysis/core/topologyattrs.py]:
```suggestion
        .. versionchanged:: 2.5.0

```

File: package/MDAnalysis/core/topologyattrs.py
Comment (hmacdope) [inline package/MDAnalysis/core/topologyattrs.py]: Are there new tests for the gyration moment itself?

**Review by hmacdope — CHANGES_REQUESTED** [review]
Looking good just a few little things :)

File: package/MDAnalysis/core/topologyattrs.py
Comment (jaclark5) [reply package/MDAnalysis/core/topologyattrs.py]: I wasn't planning on it since the shape parameter and asphericity take care of the coverage.

**Review by hmacdope — APPROVED** [review]
LGTM @jaclark5, thanks for another great contribution. @MDAnalysis/coredevs could I get a second review here?

**Comment by hmacdope** [comment]
@MDAnalysis/coredevs still looking for a second review here. :)

**Comment by hmacdope** [comment]
> @MDAnalysis/coredevs still looking for a second review here. :)

Bump on this @MDAnalysis/coredevs?

File: package/MDAnalysis/core/topologyattrs.py
Comment (orbeckst) [inline package/MDAnalysis/core/topologyattrs.py]: Why is the default for `unwrap=None` if it's a boolean? Why not `False`?

File: package/MDAnalysis/core/topologyattrs.py
Comment (orbeckst) [inline package/MDAnalysis/core/topologyattrs.py:1765]: add empty line, versionchanged is not formatted correctly

File: package/MDAnalysis/core/topologyattrs.py
Comment (orbeckst) [inline package/MDAnalysis/core/topologyattrs.py]: `(n, 3)` is a 2D array — rephrase to make clearer what you mean

File: package/MDAnalysis/core/topologyattrs.py
Comment (orbeckst) [inline package/MDAnalysis/core/topologyattrs.py]: Typeset consistently, please
* vectors with `\mathbf{r}`, tensors with `\mathsf{T}`.
* subscript labels in roman `\mathbf{r}_\mathrm{COM}` (but indices just as normal math quantities0

File: package/MDAnalysis/core/topologyattrs.py
Comment (orbeckst) [inline package/MDAnalysis/core/topologyattrs.py]: I would leave out the underscores and just call it `gyration()` or if you really want to highlight the private character then just a single underscore `_gyration()`. But double underscores are almost never needed and only make debugging harder.

File: package/MDAnalysis/core/topologyattrs.py
Comment (orbeckst) [inline package/MDAnalysis/core/topologyattrs.py]: minor PEP8 thing: space after comma

File: package/MDAnalysis/core/topologyattrs.py
Comment (orbeckst) [inline package/MDAnalysis/core/topologyattrs.py]: pep8

File: package/MDAnalysis/core/topologyattrs.py
Comment (orbeckst) [inline package/MDAnalysis/core/topologyattrs.py]: pep8

File: package/MDAnalysis/core/topologyattrs.py
Comment (orbeckst) [inline package/MDAnalysis/core/topologyattrs.py]: pep8

File: package/MDAnalysis/core/topologyattrs.py
Comment (orbeckst) [inline package/MDAnalysis/core/topologyattrs.py]: same question about `unwrap=None` ... do we do this everywhere??

File: package/MDAnalysis/core/topologyattrs.py
Comment (orbeckst) [inline package/MDAnalysis/core/topologyattrs.py]: Is this important for the user to know _how_ the quantities are computed? I'd assume the more important message is what you have in the CHANGELOG: since 2.5.0 you can calculate asphericity for any `compound`.

File: package/MDAnalysis/core/topologyattrs.py
Comment (orbeckst) [inline package/MDAnalysis/core/topologyattrs.py]: s this important for the user to know how the quantities are computed? I'd assume the more important message is what you have in the CHANGELOG: since 2.5.0 you can calculate asphericity for any `compound`.

**Review by orbeckst — CHANGES_REQUESTED** [review]
Thanks for the fix and code consolidation. I had a quick look through and mainly noticed doc things.

Did we have tests for the case where the methods actually worked as expected and do these tests still produce the same values?

File: package/MDAnalysis/core/topologyattrs.py
Comment (jaclark5) [reply package/MDAnalysis/core/topologyattrs.py]: This is not something I started. I it was the default option in `asphericity` to I followed suit.

File: package/MDAnalysis/core/topologyattrs.py
Comment (jaclark5) [reply package/MDAnalysis/core/topologyattrs.py]: This is not something I started. I it was the default option in `asphericity` so I followed suit.

File: package/MDAnalysis/core/topologyattrs.py
Comment (jaclark5) [reply package/MDAnalysis/core/topologyattrs.py:1762]: I'm not positive how to make it more clear. Since the tuple is length 2, it implies the result is 2D, and the second dimension is length 3.

File: package/MDAnalysis/core/topologyattrs.py
Comment (jaclark5) [reply package/MDAnalysis/core/topologyattrs.py]: Does the change I made clarify?

File: package/MDAnalysis/core/topologyattrs.py
Comment (jaclark5) [reply package/MDAnalysis/core/topologyattrs.py]: That's what I was going for with "for per residue quantities" but hopeful the change is more clear.

File: package/MDAnalysis/core/topologyattrs.py
Comment (orbeckst) [reply package/MDAnalysis/core/topologyattrs.py]: @richardjgowers @hmacdope do you know why we set `unwrap=None` as default instead of `unwrap=False`?

**Comment by jaclark5** [comment]
I kept the same testing structure that existed, some results stayed the same and others changed. For a calculation with the entire atom group, the result stayed the same (line 1098, 1120 in the test file.). In the tests where other `compound` options were used, the result changed as there is more than one value each reflecting a residue, while before each instance had the same value as when there was no `compound` specified.

**Comment by orbeckst** [comment]
sorry, I currently only have very limited time and won't be able to review soon

**Comment by orbeckst** [comment]
actually... that should be quick

**Review by orbeckst — APPROVED** [review]

**Comment by hmacdope** [comment]
Brilliant! two reviews. I will merge now.

**Pull request merged**
